<a href="https://colab.research.google.com/github/KARTIKPARATKAR/MACHINE-LEARNING-WORK/blob/main/DiabetesPrediction.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**Dataset Link is** - https://github.com/KARTIKPARATKAR/MACHINE-LEARNING-WORK/blob/main/diabetes.csv

**Note-**We evaluated the accuracy of the SVM model using various parameter values and kernel functions while varying the test_size.

**Problem Statement-**
We have a dataset of a patients which includes information of Pregnancies,Glucose,BloodPressure,SkinThickness,Insulin,BMI,BibetesPedigreeFunction,Age and Outcome whether the patient is diagnosed with Diabetes or not.
This is a classification problem wherein we have to apply classification ML algorithm to get output.

**Importing Dependencies-**

In [329]:
import pandas as pd
import numpy as np
from sklearn import svm
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score
from sklearn.preprocessing import StandardScaler

**Data Collection and Analysis-**

In [330]:
diabetes_dataset = pd.read_csv("/content/diabetes.csv")

In [331]:
diabetes_dataset.head()
#Here in Outcome column, 1 reresents patient is diabetic and 0 represents patient is non-diabetic.

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0            6      148             72             35        0  33.6   
1            1       85             66             29        0  26.6   
2            8      183             64              0        0  23.3   
3            1       89             66             23       94  28.1   
4            0      137             40             35      168  43.1   

   DiabetesPedigreeFunction  Age  Outcome  
0                     0.627   50        1  
1                     0.351   31        0  
2                     0.672   32        1  
3                     0.167   21        0  
4                     2.288   33        1

In [332]:
diabetes_dataset.shape
#Rows are 768 and columns are 9

(768, 9)

In [333]:
#Getting the statistical measures of the dataset
diabetes_dataset.describe()

Pregnancies     Glucose  BloodPressure  SkinThickness     Insulin  \
count   768.000000  768.000000     768.000000     768.000000  768.000000   
mean      3.845052  120.894531      69.105469      20.536458   79.799479   
std       3.369578   31.972618      19.355807      15.952218  115.244002   
min       0.000000    0.000000       0.000000       0.000000    0.000000   
25%       1.000000   99.000000      62.000000       0.000000    0.000000   
50%       3.000000  117.000000      72.000000      23.000000   30.500000   
75%       6.000000  140.250000      80.000000      32.000000  127.250000   
max      17.000000  199.000000     122.000000      99.000000  846.000000   

              BMI  DiabetesPedigreeFunction         Age     Outcome  
count  768.000000                768.000000  768.000000  768.000000  
mean    31.992578                  0.471876   33.240885    0.348958  
std      7.884160                  0.331329   11.760232    0.476951  
min      0.000000                  0.078000   21.000000    0.000000  
25%     27.300000                  0.243750   24.000000    0.000000  
50%     32.000000                  0.372500   29.000000    0.000000  
75%     36.600000                  0.626250   41.000000    1.000000  
max     67.100000                  2.420000   81.000000    1.000000

In [334]:
#Check how many diabetic datapoints and non-diabetic datapoints
diabetes_dataset['Outcome'].value_counts()

Outcome
0    500
1    268
Name: count, dtype: int64

0 --->  Non-Diabetic

1 ---> Diabetic

In [335]:
diabetes_dataset.groupby('Outcome').mean()
#This will give mean value of each column for diabetic datapoints and non-diabetic datapoints

Pregnancies     Glucose  BloodPressure  SkinThickness     Insulin  \
Outcome                                                                      
0           3.298000  109.980000      68.184000      19.664000   68.792000   
1           4.865672  141.257463      70.824627      22.164179  100.335821   

               BMI  DiabetesPedigreeFunction        Age  
Outcome                                                  
0        30.304200                  0.429734  31.190000  
1        35.142537                  0.550500  37.067164

In [336]:
x = diabetes_dataset.drop(columns = 'Outcome' , axis = 1)
y = diabetes_dataset['Outcome']

#x will store only features by dropping the outcome column
#y will store the outcome column

In [337]:
print(x)  #Data without outcome

     Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0              6      148             72             35        0  33.6   
1              1       85             66             29        0  26.6   
2              8      183             64              0        0  23.3   
3              1       89             66             23       94  28.1   
4              0      137             40             35      168  43.1   
..           ...      ...            ...            ...      ...   ...   
763           10      101             76             48      180  32.9   
764            2      122             70             27        0  36.8   
765            5      121             72             23      112  26.2   
766            1      126             60              0        0  30.1   
767            1       93             70             31        0  30.4   

     DiabetesPedigreeFunction  Age  
0                       0.627   50  
1                       0.351   31  


In [338]:
print(y)  #Data with only outcome

0      1
1      0
2      1
3      0
4      1
      ..
763    0
764    0
765    0
766    1
767    0
Name: Outcome, Length: 768, dtype: int64


We observe that all the column values are in different range. It will be difficult for our ML-model to train on a dataset which not in a perticular range. To make a data in perticular range we will use standardscalar function which is a feature scaling technique.**StandardScalar function transform data to mean value of 0 and standard-deviation value of 1**

In [339]:
scalar = StandardScaler()

In [340]:
scalar.fit(x)

StandardScaler()

In [341]:
standardized_data = scalar.transform(x)

In [342]:
print(standardized_data)  #Here we observe that all the data is in a similar range

[[ 0.63994726  0.84832379  0.14964075 ...  0.20401277  0.46849198
   1.4259954 ]
 [-0.84488505 -1.12339636 -0.16054575 ... -0.68442195 -0.36506078
  -0.19067191]
 [ 1.23388019  1.94372388 -0.26394125 ... -1.10325546  0.60439732
  -0.10558415]
 ...
 [ 0.3429808   0.00330087  0.14964075 ... -0.73518964 -0.68519336
  -0.27575966]
 [-0.84488505  0.1597866  -0.47073225 ... -0.24020459 -0.37110101
   1.17073215]
 [-0.84488505 -0.8730192   0.04624525 ... -0.20212881 -0.47378505
  -0.87137393]]


In [343]:
x = standardized_data
y = diabetes_dataset['Outcome']

In [344]:
print(x)

[[ 0.63994726  0.84832379  0.14964075 ...  0.20401277  0.46849198
   1.4259954 ]
 [-0.84488505 -1.12339636 -0.16054575 ... -0.68442195 -0.36506078
  -0.19067191]
 [ 1.23388019  1.94372388 -0.26394125 ... -1.10325546  0.60439732
  -0.10558415]
 ...
 [ 0.3429808   0.00330087  0.14964075 ... -0.73518964 -0.68519336
  -0.27575966]
 [-0.84488505  0.1597866  -0.47073225 ... -0.24020459 -0.37110101
   1.17073215]
 [-0.84488505 -0.8730192   0.04624525 ... -0.20212881 -0.47378505
  -0.87137393]]


In [345]:
print(y)

0      1
1      0
2      1
3      0
4      1
      ..
763    0
764    0
765    0
766    1
767    0
Name: Outcome, Length: 768, dtype: int64


**Splitting data in training data and test data**

In [346]:
x_train , x_test , y_train , y_test = train_test_split(x,y,test_size=0.25 , stratify = y , random_state = 2)
#stratify = y means that our data will be splitted equally on the basis of diabetic datapoints and non-diabetic datapoints.

In [347]:
print(x.shape , x_train.shape , x_test.shape)

(768, 8) (576, 8) (192, 8)


**Training the SVM Model-**


In [348]:
svm_model = svm.SVC(kernel='sigmoid')

In [349]:
svm_model.fit(x_train,y_train)

SVC(kernel='sigmoid')

**Model Evaluation-**

In [350]:
#Accuracy score on training data
x_train_accuracy = svm_model.predict(x_train)
training_data_accuracy = accuracy_score(x_train_accuracy,y_train)

In [351]:
print("Accuracy Score of Training data-",training_data_accuracy)

Accuracy Score of Training data- 0.7013888888888888


In [352]:
#Accuracy Score on testing data
x_test_accuracy = svm_model.predict(x_test)
testing_data_accuracy = accuracy_score(x_test_accuracy,y_test)

In [353]:
print("Accuracy Score on Testing data-",testing_data_accuracy)

Accuracy Score on Testing data- 0.6822916666666666


**Building a predictive system-**

In [354]:
input_data = (5,109,75,26,0,36,0.546,60)
#Changing input data as numpy array
input_data_as_numpy_array = np.asarray(input_data)

#Reshaping the array as we are predicting for one instance and not for all the datapoints available
input_data_reshaped = input_data_as_numpy_array.reshape(1,-1)


#When we trained the model,we standardized the data then trained the model so
#when we will predict the output by predictive system , we will need the standardized data only.
#So we will have to stndardized this input data.
std_data = scalar.transform(input_data_reshaped)

print(std_data)

prediction = svm_model.predict(std_data)
print(prediction)

if prediction[0] == 0 :
  print("The person is not diabetic")
else :
  print("The person is diabetic")


[[ 0.3429808  -0.37226487  0.304734    0.34271738 -0.69289057  0.50861896
   0.22386236  2.27687294]]
[0]
The person is not diabetic


/usr/local/lib/python3.11/dist-packages/sklearn/utils/validation.py:2739: UserWarning: X does not have valid feature names, but StandardScaler was fitted with feature names
  warnings.warn(


**All the following accuracies are observed when we kept test_size =0.2**

accuracy = 0.7727272727272727 for linear kernel

accuracy =  0.7077922077922078 for poly kernel with degree 3

accuracy =  0.6948051948051948 for poly kernel with degree 2

accuracy = 0.7272727272727273 for rbf kernel

accuracy = 0.7532467532467533 for sigmoid kernerl


**All the following accuracies are observed when we kept test_size =0.25**

accuracy = 0.765625 for linear kernel

accuracy =  0.7291666666666666 for poly kernel with degree 3

accuracy =  0.6979166666666666 for poly kernel with degree 2

accuracy = 0.7395833333333334 for rbf kernel

accuracy = 0.6822916666666666 for sigmoid kernerl
